python/tests/param/params_test.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2017 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import, division, print_function
+
+from collections import namedtuple
+from six import with_metaclass
+
+from sparkdl.param.converters import SparkDLTypeConverters
+
+from ..tests import PythonUnitTestCase
+
+
+class TestGenMeta(type):
```

Comment:
@phi-dbq sorry, this file is beyond my cognitive load to comprehend.

If I am having a hard time to understand it now, I will definitely have an even harder time to diagnose why a test a failing. Please write down every test separately using a function. A test has multiple purposes, in order of priorities:
  1. provide through some examples the specification of a function
  2. when a modification is done subsequently, provide some assistance in diagnosing the failure
  3. and least important, test a specific implementation

This current file is probably right on the last point, but I am not able to verify the first point, and I am certain it will be not be of assistance for the second point. I am happy to discuss it more if you want.

Reply:
I updated the test cases by using [parameterized](https://github.com/wolever/parameterized)  to generate individual test functions.

Reply:
Adding methods to a test class programmatically is a must for our use case. The unit-test framework we use does not have good support for parameterized tests, leaving us no choice but to look externally.

Meta-classes are a supported feature in python [from Guido's treatise](https://www.python.org/download/releases/2.2.3/descrintro/#metaclasses).
In fact, we use it already [in this codebase](https://github.com/databricks/spark-deep-learning/blob/master/python/sparkdl/transformers/keras_applications.py#L37).
It might require some effort to parse, but they are natively supported, making the test infra easier to maintain.

On the other hand, using an external project might make code look cleaner, but it might also add considerable maintenance burden to us. Fortunately, the one we found, [parameterized](https://github.com/wolever/parameterized), is an actively maintained small project.